### ISSUE TITLE
Query sometimes returns duplicate documents

### ISSUE THREAD

**Issue by bjoerge** [sanity-io/gradient#631]
This seems to be an issue only with large result sets.

Here's a reproducible test case (might need a couple of runs for it to show duplicates):

```sh
curl 'https://ppsg7ml5.api.sanity.io/v1/data/query/test?query=*%5B_type%20%3D%3D%20%24type%5D%20%7C%20order(_updatedAt%20desc)%20%5B0...10000%5D%20%7B_id%2C%20_type%7D&%24type=%22species%22&%24draftsPath=%22drafts.**%22'\
 -H 'accept-encoding: gzip, deflate, br'\
 -H 'accept: application/json'\
 --compressed\
 | jq '.result | group_by(._id) | map(.[]+{"dupes":length}) | unique_by(._id) | map(select(.dupes > 1))'
```

**Comment by bjoerge** [comment]
For some reason, I'm no longer able to reproduce this on the `test` dataset above. If I switch to another dataset (`species`), I get duplicates after a few attempts:

```sh
curl 'https://ppsg7ml5.api.sanity.io/v1/data/query/species?query=*%5B_type%20%3D%3D%20%24type%5D%20%7C%20order(_updatedAt%20desc)%20%5B0...10000%5D%20%7B_id%2C%20_type%7D&%24type=%22species%22&%24draftsPath=%22drafts.**%22'\
 -H 'accept-encoding: gzip, deflate, br'\
 -H 'accept: application/json'\
 --compressed\
 | jq '.result | group_by(._id) | map(.[]+{"dupes":length}) | unique_by(._id) | map(select(.dupes > 1))'
```

(I'll leave this dataset alone for debugging now)

**Comment by bjoerge** [comment]
Side note: I've never seen more than two duplicates of the same document.

**Comment by erikgrinaker** [comment]
Tried fetching the documents directly from ES (via `_search`), this consistently gave 0 duplicates against every cluster node.

**Comment by erikgrinaker** [comment]
Also, when running the query with `0...1000` it sometimes returns 0 matches for e.g. `_id: "tremaster-mirabilis"`, but other times 1 or 2 matches.

**Comment by erikgrinaker** [comment]
I think I've found the problem here.

We run ElasticSearch on three nodes in production (`b`, `c`, and `d`), with 3 shards and 1 replica per index (6 shards total). When querying each ElasticSearch node directly, it returns equivalent results when fetching all documents (i.e. `"from": 0, "size": 10000`), but the order is different. The reason is that most of the documents have identical sort keys (i.e. `_updatedAt`), and ordering within the same sort key is undefined. This order will typically depend on the internal segment structure of the underlying Lucene indices, which varies between shards/nodes due to different write patterns.

This becomes a problem due to how Gradient does ElasticSearch pagination. Gradient first fetches pages of 100 results from 0 to 1000 individually, and only then switches to a scroll-based API to fetch the remainder. However, each of the first 100-result pages is fetched in a separate HTTP request, thus each page may be fetched from a different node/shard depending on load balancing. Since the sort order is different between nodes, page 2 from node `c` will not contain the same results as page 2 from node `d`. Thus, the final result may contain the same document from page 2 on node `b`, and from page 3 on node `c`, yielding duplicates and/or missing documents.

@atombender A naive solution would be to fetch the entire result in a single query if the query results fall within the first 1000 documents, and always use scrolling otherwise. However, this will cause inconsistencies with our own pagination - e.g. queries for `[0...100]` followed by `[100...200]` may contain duplicates depending on which ES node happens to process the request. Maybe we should include our own UUID attr on every document, and implicitly order on this to preserve a deterministic order across all shards?

**Comment by erikgrinaker** [comment]
On second thought, we should just implicitly order on `_id` as a tie-breaker, since this is guaranteed to be unique.

**Comment by erikgrinaker** [comment]
Implicit ordering on `_id` will have serious performance ramifications, since even queries where the user does not specify an order will have to be ordered to be consistent across pages. Perhaps we should deprecate pagination, and instead implement proper cursor support with iterators in the clients?

**Comment by erikgrinaker** [comment]
Turns out we already used a deterministic default order, it just wasn't included as a tie-breaker when the query specifies an ordering. Fix is in #633.

**Comment by atombender** [comment]
The reason we transition from a paginated query to a scrolled one is that pagination is very inefficient beyond a certain offset, whereas scrolling is expensive on the ES side, as it's stateful; the scroll ID is associated with a node that has to keep an internal index reader open.

But this means we _have_ to do scrolling internally if we are to allow people to fetch more than, say, 1,000 documents or whatever we set the threshold to be.

A cursor API would be better, but that will require some more design, and doesn't necessarily avoid the problem. For example, a single GROQ query can spawn multiple join queries, and dependent on whether they return a lot of data, the system still needs to decide whether to use pagination or scrolling there.

Knowing how many documents the ES query will need to access should be possible, but it probably requires flowing more metadata through the pipeline that we do today. However, we push the limit down into ES whenever possible, so a decent temporary solution would be to always use scrolling if either we don't have a limit, or the limit is larger than, say, 1,000.
